File: daily_problems/2026/06/0626/personal_submission/cf106020c_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"
#include "atcoder/segtree"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

long long op(long long x, long long y) {return gcd(x, y);}

long long e() {return 0;}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<long long> nums(n);
	for (auto &x: nums) cin >> x;

	vector<int> deg(n, 0);
	for (int i = 0; i < n - 1; i ++) {
		int u, v;
		cin >> u >> v;
		u --, v --;
		deg[u] ++, deg[v] ++;
	}

	vector<int> tmp;
	for (int i = 0; i < n; i ++) if (deg[i] == 1) tmp.emplace_back(i);
	for (int i = 0; i < n; i ++) if (deg[i] > 1) tmp.emplace_back(i);

	vector<long long> vals(n, 0);
	vector<int> pos(n, 0);

	for (int i = 0; i < n; i ++) {
		vals[i] = nums[tmp[i]];
		pos[tmp[i]] = i;
	}

	atcoder::segtree<long long, op, e> seg(vals);

	int q;
	cin >> q;

	while (q --) {
		int idx;
		long long val;
		cin >> idx >> val; idx --;

		vals[pos[idx]] = val;
		seg.set(pos[idx], val);

		int k = min(16, n);
		long long cur = seg.prod(k, n);

		vector<long long> pref = {0};
		for (int i = 0; i < k - 1; i ++) pref.emplace_back(gcd(pref.back(), vals[i]));

		bool flg = false;
		for (int i = k - 1; i >= 0; i --) {
			if (deg[tmp[i]] == 1 && gcd(pref[i], cur) == 1) flg = true;
			cur = gcd(cur, vals[i]);
		}

		cout << (flg ? "Yes" : "No") << '\n';
	}

	return 0;
}

File: daily_problems/2026/06/0626/personal_submission/cf106020c_yawn_sean.py
# Submission link: https://codeforces.com/gym/106020/submission/380104879
def main():
    n = II()
    nums = LII()
    
    path = [[] for _ in range(n)]
    for _ in range(n - 1):
        u, v = GMI()
        path[u].append(v)
        path[v].append(u)
    
    tmp = []
    for i in range(n):
        if len(path[i]) == 1:
            tmp.append(i)
    
    for i in range(n):
        if len(path[i]) > 1:
            tmp.append(i)
    
    vals = [0] * n
    pos = [0] * n
    
    for i in range(n):
        vals[i] = nums[tmp[i]]
        pos[tmp[i]] = i
    
    seg = SegTree(math.gcd, 0, vals)
    
    q = II()
    outs = []
    
    for _ in range(q):
        i, x = MII()
        i -= 1
        
        vals[pos[i]] = x
        seg.set(pos[i], x)
        
        k = fmin(16, n)
        val = seg.prod(k, n)
        
        pref = [0]
        for i in range(k - 1):
            pref.append(math.gcd(pref[-1], vals[i]))
        
        flg = False
        for i in range(k - 1, -1, -1):
            if len(path[tmp[i]]) == 1 and math.gcd(pref[i], val) == 1:
                flg = True
            val = math.gcd(vals[i], val)
        
        outs.append('Yes' if flg else 'No')
    
    print('\n'.join(outs))